include/nbl/builtin/hlsl/memory_accessor.hlsl

```diff
@@ -16,73 +16,43 @@ struct MemoryAdaptor
 {
     BaseAccessor accessor;
     
-    // TODO: template all get,set, atomic... then add static_asserts of `has_method<BaseAccessor,signature>::value`, do vectors and matrices in terms of each other
-    uint get(const uint ix) { return accessor.get(ix); }
-    void get(const uint ix, NBL_REF_ARG(uint) value) { value = accessor.get(ix);}
-    void get(const uint ix, NBL_REF_ARG(uint2) value) { value = uint2(accessor.get(ix), accessor.get(ix + _NBL_HLSL_WORKGROUP_SIZE_));}
-    void get(const uint ix, NBL_REF_ARG(uint3) value) { value = uint3(accessor.get(ix), accessor.get(ix + _NBL_HLSL_WORKGROUP_SIZE_), accessor.get(ix + 2 * _NBL_HLSL_WORKGROUP_SIZE_));}
-    void get(const uint ix, NBL_REF_ARG(uint4) value) { value = uint4(accessor.get(ix), accessor.get(ix + _NBL_HLSL_WORKGROUP_SIZE_), accessor.get(ix + 2 * _NBL_HLSL_WORKGROUP_SIZE_), accessor.get(ix + 3 * _NBL_HLSL_WORKGROUP_SIZE_));}
-
-    void get(const uint ix, NBL_REF_ARG(int) value) { value = asint(accessor.get(ix));}
-    void get(const uint ix, NBL_REF_ARG(int2) value) { value = asint(uint2(accessor.get(ix), accessor.get(ix + _NBL_HLSL_WORKGROUP_SIZE_)));}
-    void get(const uint ix, NBL_REF_ARG(int3) value) { value = asint(uint3(accessor.get(ix), accessor.get(ix + _NBL_HLSL_WORKGROUP_SIZE_), accessor.get(ix + 2 * _NBL_HLSL_WORKGROUP_SIZE_)));}
-    void get(const uint ix, NBL_REF_ARG(int4) value) { value = asint(uint4(accessor.get(ix), accessor.get(ix + _NBL_HLSL_WORKGROUP_SIZE_), accessor.get(ix + 2 * _NBL_HLSL_WORKGROUP_SIZE_), accessor.get(ix + 3 * _NBL_HLSL_WORKGROUP_SIZE_)));}
+    // TODO: template atomic... then add static_asserts of `has_method<BaseAccessor,signature>::value`, do vectors and matrices in terms of each other
+    uint get(const uint ix) 
+    { 
+        uint retVal;
+        accessor.get(ix, retVal);
+        return retVal; 
+    }
 
-    void get(const uint ix, NBL_REF_ARG(float) value) { value = asfloat(accessor.get(ix));}
-    void get(const uint ix, NBL_REF_ARG(float2) value) { value = asfloat(uint2(accessor.get(ix), accessor.get(ix + _NBL_HLSL_WORKGROUP_SIZE_)));}
-    void get(const uint ix, NBL_REF_ARG(float3) value) { value = asfloat(uint3(accessor.get(ix), accessor.get(ix + _NBL_HLSL_WORKGROUP_SIZE_), accessor.get(ix + 2 * _NBL_HLSL_WORKGROUP_SIZE_)));}
-    void get(const uint ix, NBL_REF_ARG(float4) value) { value = asfloat(uint4(accessor.get(ix), accessor.get(ix + _NBL_HLSL_WORKGROUP_SIZE_), accessor.get(ix + 2 * _NBL_HLSL_WORKGROUP_SIZE_), accessor.get(ix + 3 * _NBL_HLSL_WORKGROUP_SIZE_)));}
+    template<typename Scalar>
+    void get(const uint ix, NBL_REF_ARG(Scalar) value) { accessor.get(ix, value);}
```

Comment:
HLSL is weird and allows for `inout` to do conversions when copying in and copying out, this won't protect you against `get<float>(const uint, inout float)` calling into `accessor.get(const uint, inout uint)` and doing a horrible conversion on the way

Reply:
if you want to do it this way, you need an `impl::MemoryAdaptor_get<BaseAccessor,Scalar>` struct with `void get(NBL_REF_ARG(BaseAccessor) accessor, NBL_REF_ARG(Scalar) value)` and `set` methods.

then partial specializations of that struct so you have `floatN_t` and `intN_t` done in terms `get/set` of `uintN_t` with a `bitcast`

Reply:
Did it different, lmk if what's currently up is fine

Reply:
looks fine, resolved.